### PR TITLE
Added DoublePulsar detection to MS17-010 scanner module

### PR DESCRIPTION
**Pull request by ghost**
Adds DoublePulsar detection to my MS17-010 scanner module landed in #8167.  This performs the DoublePulsar SMB trans2 knock with MultiplexID 0x41, and if it knocks back with MultiplexID 0x51 it's an indicator of pre-existing compromise. DoublePulsar lives as shellcode in ring0 and goes away on system restarts, but seems to change this multiplexID away from the SMB specification. Others have noticed if you knock with 0x42, you'll get 0x52, so it's adding 0x10 to the MID when it should be the same.

> The client insures that every request that is sent to the server has a unique MID.  When the server replies to a given request, the response contains the same MID.  In this way, multiple requests can be sent to the server, and the client can simply match the response MID with the MID it generated to know which request has just been replied to.

This knock is as non-intrusive and unauthed  as the missing patch information disclosure, and is done during the same SMB connection if the patch is missing (it's simple to check for, I don't think it needs to be an additional toggle option or a separate module, but will go that route if it's wanted). 

I also cleaned up some code and all print statements are now in main run function. I inversed the print messages from good<->bad to better reflect the rest of the framework as a red team tool. DoublePulsar detection is a warning message.

Code is based on Luke Jennings Python code (credited in module)

https://github.com/countercept/doublepulsar-detection-script

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] `use auxiliary/scanner/smb/smb_ms17_010`
- [x] target host missing ms17-010
- [ ] **Verify** module says host is vulnerable, no message about infection
- [ ] hook host with DoublePulsar
- [ ] **Verify** module says host is vulnerable AND infected 
- [ ] **Verify** the "XOR Key" in module output matches DoublePulsar ping

### CODE REVIEW

**Comment by wvu-r7** [comment]
Thanks. This is better than a separate module.

**Comment by jork2345** [comment]
waiting for an exploit module so hard ! ;)

**Comment by zerosum0x0** [comment]
I figured out the DoublePulsar "xor key" algorithm and think it should be included in this module.

s = SMB.Signature
x = 2 * s ^ (((s & 0xFF00 | (s << 16)) << 8) | (((s >> 16) | s & 0xFF0000u) >> 8));

**Comment by zerosum0x0** [comment]
Added the XOR key, changed pull request verification steps to verify it

**Comment by wvu-r7** [comment]
Nice work, @zerosum0x0. I was looking at the XOR key algorithm just a couple days ago, too. I've been toying with the idea of a DOUBLEPULSAR exploit module, but I haven't gotten around to writing anything up yet. If you'd like to do it, please feel free to. You have a better understanding of the code than I do.

**Comment by zerosum0x0** [comment]
I probably will hack up something rudimentary in the next few days. But I would definitely need some help from you and others in organizing it since you are more familiar with the framework.

At first we can just make a module with a lot of hardcoded stuff that interacts with DoublePulsar. One thing that needs to get separated out and cleaned up is a raw kernel payload to inject DLL. Perhaps two, one that takes an existing process name (such as lsass.exe [i.e. what DP does]) and one which spawns a hidden process with a system token. I don't know if Metasploit has such a thing already?

I've also pretty much reversed all of pertinent parts of EternalBlue.exe. One thing I'd like to avoid is a lot of duplicated effort (such as what happened with EXTRABACON, which was my fault). I don't know if anyone else is planning on porting it to Metasploit.

File: modules/auxiliary/scanner/smb/smb_ms17_010.rb [hunk 1]
```diff
@@ -35,27 +42,55 @@ def initialize(info = {})
...
+
+      tree_id = do_smb_setup_tree(ipc_share)
+      print_status("Connected to #{ipc_share} with TID = #{tree_id}")
+
+      status = do_smb_ms17_010_probe(tree_id)
+      print_status("Received #{status} with FID = 0")
 
       if status == "STATUS_INSUFF_SERVER_RESOURCES"
-        print_warning("Host is likely VULNERABLE to MS17-010!")
+        print_good("Host is likely VULNERABLE to MS17-010!")
```
Comment (timwr) [inline modules/auxiliary/scanner/smb/smb_ms17_010.rb]: Good change

**Review by timwr — APPROVED** [review]
Looks good to me. Excellent stuff @zerosum0x0

File: modules/auxiliary/scanner/smb/smb_ms17_010.rb (hunk 1 above)
Comment (wvu-r7) [inline modules/auxiliary/scanner/smb/smb_ms17_010.rb]: :+1:

**Comment by zerosum0x0** [comment]
I added printing simple.client.peer_native_os for vulnerable hosts, so you don't waste time trying to pop Windows 10 boxes with EternalBlue.

This can probably be pulled if it passes all tests. I will refactor some code out if needed in a future pull request.

**Comment by nixawk** [comment]
A python script for SMB is created. https://github.com/nixawk/labs/blob/master/MS17_010/smb_exploit.py

**Comment by zerosum0x0** [comment]
@nixawk nice. I actually started this module doing something similar in Python, but went the lazy route with Metasploit since it had a decent (and now better) SMB library.

You should add DP detection and print the OS like my script (as well as support RHOSTS)

**Comment by wvu-r7** [comment]
@zerosum0x0: Sorry for the delayed response.

1. Feel free to create a mixin for now. As I understand it, the Trans2 packet is being worked on by @dmaloney-r7, and we should be able to use that when it's ready.
2. Don't think we have any kernel payloads, much less kernel-based DLL injection. I've been injecting into `spoolsv.exe`.
3. I wanted to say `regenerate_payload` could do it, but if you choose a payload with a specific arch (which is usually the case for native payloads), you're tied to that arch.
4. All yours! We've been doing our own analysis, but we've made no movement on a module. I know you'll do it justice. :)

Let me know when you'd like this PR merged. Thanks!

**Comment by zerosum0x0** [comment]
thanks @wvu-r7  You can land this for now, so that people who clone can get it as-is. Serving us real well on pentests with the OS detection.

**Comment by zerosum0x0** [comment]
@wvu-r7 spoolsv.exe much saner than the default lsass.exe! good tip!

**Comment by wvu-r7** [comment]
Shipped, yo. Thanks for the badass research. It's definitely filled in the gaps in my understanding. Cheers!

**Comment by nixawk** [comment]
Thanks @zerosum0x0. I've add DP detection / Native OS feature. Please try it against your lab.

**Comment by wvu-r7** [comment]
# Release Notes

This adds DOUBLEPULSAR detection to the MS17-010 scanner module.

**Pull request merged**